### ISSUE TITLE
The stuttering of ComposeUI on iOS Device

### ISSUE THREAD

**Issue by cx478815108** [JetBrains/compose-jb#2283]
Hello，I am using compose-jb. I wrote a long list page with LazyColumn and found some performance issue, mostly stuttering. 

In order to determine the specific cause of the stuttering, I did some experiments, and with the help of Xcode Instruments, I eliminated some interference and determined the main cause.

When most of the stuck occurs, it is mainly caused by the untimely callback of FrameDispatcher onFrame on iOS.

I'm not sure whether it is caused by ComposeUI's untimely dispatch of tasks or the scheduling of coroutines.

Are there any plans here to optimize stuttering on iOS?

Looking forward to reply

**Comment by yangwuan55** [comment]
Supported?for iOS?

**Comment by pjBooms** [comment]
We need a reproducing example and probably a written script of your performance analyses to continue working on this issue

**Comment by cx478815108** [comment]
> We need a reproducing example and probably a written script of your performance analyses to continue working on this issue

yeah I have prepared a demo project. if you run the project when using iPhone7 or some lower performance iPhone. 
You can see frame drops visible to the eye

**Comment by pjBooms** [comment]
@cx478815108 Finally, we have successfully launched your example. Unfortunately, we cannot notice obvious sluttering. 
What iOS device do you use for testing? Do you test in Release mode?
Can you record a video to show the sluttering on your side?

**Comment by mahozad** [comment]
Is the following an example of stuttering?

https://user-images.githubusercontent.com/29678011/206242186-5876e1a1-84a5-4259-8065-56b3630d7f5e.mp4

The app prints `Failed to create DirectX12 device.` on start.

### Relatd issues:
- #610
- https://github.com/JetBrains/skiko/issues/601
- #553
- #446
- #442

### This is happening for me on Desktop:
- OS: Windows 11 Version 21H2 Build 22000.1281
- Integrated graphics card: Intel® HD Graphics 4600
- Dedicated graphics card: AMD Radeon R7 M265
- Compose Multiplatform: 1.2.1
- Java: Oracle OpenJDK 17.0.5
- Kotlin: 1.7.20

### Minimal reproducible project: 
[compose-stutter.zip](https://github.com/JetBrains/compose-jb/files/10178486/compose-stutter.zip)

**Comment by pjBooms** [comment]
>>Is the following an example of stuttering?
Thank you! We will look into it

**Comment by cx478815108** [comment]
> What iOS device do you use for testing

I am using iPhone7

**Comment by pjBooms** [comment]
That is super cool! Thank you very much.

>The first is that LazyColumn nests LazyRow, and there is a stack below.
Do you have .json profile for this too?

I have started to analyze https://github.com/cx478815108/ListSample/blob/main/stuckrecord/stutterOfGetAndMeasure.json profile.

Currently it seems to be K/N bound, but anyway we need to get a closer look what we can do on Compose level to optimize it.

**Comment by pjBooms** [comment]
May you share your TMMSignPost class? (I believe it is os_signpost wrapper) 
@cx478815108

**Comment by cx478815108** [comment]
> Currently it seems to be K/N bound, but anyway we need to get a closer look what we can do on Compose level to optimize it.

I am a senior iOS developer. Through my own efforts, I have mastered some points that need to be optimized, as follows:
1. LazyColumn nested LazyRow's Caton
2. Compose rasterizes on the main thread, which requires further optimization
3. For example, the dispatch function of Dispatchers.Main

I recorded more points, which can be optimized from the Compose level.
If possible, can I participate in the optimization together?

**Comment by cx478815108** [comment]
>

**Comment by pjBooms** [comment]
>Our team also fixed the crash of Compose on iOS 11-12.

**Comment by pjBooms** [comment]
>I believe that Google as well should be involved in this process.

**Comment by cx478815108** [comment]
> Please email me on the subject.

<img width="285" alt="image" src="https://user-images.githubusercontent.com/9649099/207765620-123d1974-d023-42f7-9c04-524471297f32.png">
I have sent you an email, thank you very much.

**Comment by okushnikov** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/CMP-2283">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
